thewalrus/_torontonian.py

```diff
@@ -54,3 +54,214 @@ def tor(A, fsum=False):
+
+
+    Args:
+        mu (1d array) : means of xp Gaussian Wigner function
+        cov (2d array) : : xp Wigner covariance matrix
+        det_pattern (1d numpy array) : array of {0,1} to describe the threshold detection outcome
+        hbar (float): the value of :math:`\hbar` in the commutation relation :math:`[\x,\p]=i\hbar`.
+
+    Returns:
+        np.float64 : probability of detection pattern
+    """
+    det_pattern = np.asarray(det_pattern)
+
+    m = len(cov)
+    assert cov.shape == (m, m)
+    assert m % 2 == 0
+    n = m // 2
+
+    means_x = mu[:n]
+    means_p = mu[n:]
+    avec = np.concatenate((means_x + 1j * means_p, means_x - 1j * means_p), axis=0) / np.sqrt(2 * hbar)
+
+    Q = Qmat_numba(cov, hbar=hbar)
+
+    if max(det_pattern) > 1:
+        raise ValueError(
+            "When using threshold detectors, the detection pattern can contain only 1s or 0s.")
+
+    nonzero_idxs = np.where(det_pattern == 1)[0]
+    zero_idxs = np.where(det_pattern == 0)[0]
+
+    ii1 = np.concatenate((nonzero_idxs, nonzero_idxs + n), axis=0)
+    ii0 = np.concatenate((zero_idxs, zero_idxs + n), axis=0)
+
+    Qaa = numba_ix(Q, ii0, ii0)
+    Qab = numba_ix(Q, ii0, ii1)
+    Qba = numba_ix(Q, ii1, ii0)
+    Qbb = numba_ix(Q, ii1, ii1)
+
+    Qaa_inv = np.linalg.inv(Qaa)
+    Qcond = Qbb - Qba @ Qaa_inv @ Qab
+
+    avec_a = avec[ii0]
+    avec_b = avec[ii1]
+    avec_cond = avec_b - Qba @ Qaa_inv @ avec_a
+
+    p0a_fact_exp = np.exp(avec_a @ Qaa_inv @ avec_a.conj() * (-0.5)).real
+    p0a_fact_det = np.sqrt(np.linalg.det(Qaa).real)
+    p0a = p0a_fact_exp / p0a_fact_det
+
+    n_det = len(nonzero_idxs)
+    p_sum = 1.  # empty set is not included in the powerset function so we start at 1
+    for z in powerset(np.arange(n_det)):
+        Z = np.asarray(z)
+        ZZ = np.concatenate((Z, Z + n_det), axis=0)
+
+        avec0 = avec_cond[ZZ]
+        Q0 = numba_ix(Qcond, ZZ, ZZ)
+        Q0inv = np.linalg.inv(Q0)
+
+        fact_exp = np.exp(avec0 @ Q0inv @ avec0.conj() * (-0.5)).real
+        fact_det = np.sqrt(np.linalg.det(Q0).real)
+
+        p_sum += ((-1) ** len(Z)) * fact_exp / fact_det
+
+    return p0a * p_sum
+
+def threshold_detection_prob(mu, cov, det_pattern, hbar=2, tol=1e-10): 
+    r"""Threshold detection probabilities for Gaussian states.
+    Formula from Jake Bulmer and Stefano Paesani.
+    When state is displaced, threshold_detection_prob_displacement is called.
+    Otherwise, tor is called.
+
+    Args:
+        mu (1d array) : means of xp Gaussian Wigner function
+        cov (2d array) : : xp Wigner covariance matrix
+        det_pattern (1d array) : array of {0,1} to describe the threshold detection outcome
+        hbar (float): the value of :math:`\hbar` in the commutation relation :math:`[\x,\p]=i\hbar`.
+        tol (float): tolerance for determining if displacement is negligible
+
+    Returns:
+        np.float64 : probability of detection pattern
+    """
+
+    if np.linalg.norm(mu) < tol:

```

Comment:
It would be better to compare using `np.allclose` and get `atol` and `rtol` in the signature of the function.

Reply:
okay :) 

btw I am still working on this. I was having a bit of trouble with my build environments, which is why I've commited a bunch of code which doesn't really work! Currently still trying to get the tests to pass

Reply:
Ups, maybe I am too eager. Just leave a message in the PR when you want to have it reviewed. One final that I forgot to mention: we should add an `hbar` parameter to all the tests being added here to make sure that everything works even if the user does not want to use the usual convention.